src/gululog_idx.erl

```diff
@@ -234,6 +235,27 @@ delete_from_cache(Tid, LogId) ->
   end,
   Tid.
 
+%% @doc Truncate after given logid from cache and index file
+%% Return new index and index file for deleted or truncated
+%% @End
+-spec truncate(dirname(), index(), segid(), logid(), ?undef | dirname()) ->
+        {index(), [filename()]}.
+truncate(Dir, #idx{tid = Tid, fd = Fd} = Idx, SegId, LogId, BackupDir) ->
+  Ms = ets:fun2ms(fun(?ETS_ENTRY(_, LogIdX, _) = EtsEntry) when LogIdX >= LogId -> EtsEntry end),
+  [_ | _] = EtsEntryList = ets:select(Tid, Ms),
+  lists:map(fun(?ETS_ENTRY(_, LogIdX, _)) -> ets:delete(Tid, LogIdX) end, EtsEntryList),
+  DeleteList = lists:filter(fun(?ETS_ENTRY(SegIdX, _, _)) -> SegIdX > SegId end, EtsEntryList),
+  %% close writer fd
+  file_sync_close(Fd),
+  %% delete idx file for > segid
+  DeleteResult = truncate_delete_do(DeleteList, Dir, BackupDir),
+  %% truncate idx file for = segid
+  TruncateResult = truncate_truncate_do(Dir, SegId, LogId, BackupDir),
+  %% re-open writer fd
+  [_ | _] = NewIndexFiles = wildcard_reverse(Dir),
+  {NewVersion, NewWriterFd} = open_writer_fd(false, hd(NewIndexFiles)),
```

Comment:
gululog_topic should perform a switch right after the truncation in case the version in the latest segment after truncation is < ?LOGVSN.
here comes a tricky case, if truncate from the verify first logid in a segment, it will be only a version byte left in the file after truncation, then gululog_topic will switch, then the file is left with only one byte content.